src/session_mbedtls.c

```diff
@@ -1637,45 +1634,57 @@ nc_server_tls_parse_crl_dist_points(unsigned char **p, size_t len, char ***uris,
-                    ERR(NULL, "Failed to parse CRL distribution points extension (%s).", nc_get_mbedtls_str_err(ret));
+                ret = mbedtls_asn1_get_tag(p, end_general_names, &name_len, MBEDTLS_ASN1_CONSTRUCTED | MBEDTLS_ASN1_SEQUENCE);
+                if (ret) {
+                    ERR(NULL, "Failed to parse GeneralName in CRL distribution points extension (%s)", nc_get_mbedtls_str_err(ret));
                     goto cleanup;
                 }
 
-                if (san.type == MBEDTLS_X509_SAN_UNIFORM_RESOURCE_IDENTIFIER) {
-                    /* found an URI */
-                    tmp = realloc(*uris, (*uri_count + 1) * sizeof **uris);
-                    if (!tmp) {
-                        ERRMEM;
-                        ret = 1;
-                        mbedtls_x509_free_subject_alt_name(&san);
+                unsigned char *end_names = *p + name_len;
+                while (*p < end_names) {
+                    unsigned char tag = **p;
+                    unsigned char tag_class = tag & MBEDTLS_ASN1_TAG_CLASS_MASK;
+                    unsigned char tag_value = tag & MBEDTLS_ASN1_TAG_VALUE_MASK;
+
+                    /* Read the GeneralName tag and length */
+                    ret = mbedtls_asn1_get_tag(p, end_names, &name_len, MBEDTLS_ASN1_CONTEXT_SPECIFIC | (tag & MBEDTLS_ASN1_CONSTRUCTED) | tag_value);
+                    if (ret) {
+                        ERR(NULL, "Failed to parse GeneralName in CRL distribution points extension (%s).",
+                            nc_get_mbedtls_str_err(ret));
                         goto cleanup;
                     }
-                    *uris = tmp;
 
-                    (*uris)[*uri_count] = strndup((const char *)san.san.unstructured_name.p, san.san.unstructured_name.len);
-                    if (!(*uris)[*uri_count]) {
-                        ERRMEM;
-                        ret = 1;
-                        mbedtls_x509_free_subject_alt_name(&san);
-                        goto cleanup;
+                    if (tag_class == MBEDTLS_ASN1_CONTEXT_SPECIFIC && tag_value == 6) {
+                        /* uniformResourceIdentifier [6] IA5String */
+                        char *uri = malloc(name_len + 1);
+                        if (!uri) {
+                            ERRMEM;
+                            ret = 1;
+                            goto cleanup;
+                        }
+                        memcpy(uri, *p, name_len);
+                        uri[name_len] = '\0';
+
+                        tmp = realloc(*uris, (*uri_count + 1) * sizeof **uris);
+                        if (!tmp) {
+                            ERRMEM;
+                            free(uri);
+                            ret = 1;
+                            goto cleanup;
+                        }
+                        *uris = tmp;
+                        (*uris)[*uri_count] = uri;
+                        ++(*uri_count);
```

Comment:
I put some more thought into this one... We should probably `strndup` first because if that happened to fail, there is no need to `realloc`. Let me know what you think

Reply:
Yes, I also think doing it in this order is much better.